### PR TITLE
Farm: add option to harvest late when not unlocking

### PR DESCRIPTION
**Pull request by Daniferrito**
Also restructured condition on when to harvest berries. The new version should be much easier to follow.

The new condition should still work the same, but I have not run the tests because they are not setup to be easy to run from outside the github workflow. I might take a stab on making the tests a one-comand to run later.

### CODE REVIEW

**Comment by Daniferrito** [comment]
I was trying it with wacan berries, and with enough of them, not calculating the growth multiplier makes the berry time out instead of be collected. It should only ever be calculated if late harvesting is requested, and anyways, the base game calculated it for all plots every single frame, so this is barely noticeable.

File: src/lib/Farm.js [hunk 1]
```diff
@@ -281,34 +302,70 @@ class AutomationFarm
...
+                }
+                // If the strategy requires to harvest the berries as late as possible, wait until the berry is about to die
+                if ((this.__internal__currentStrategy.harvestStrategy === this.__internal__harvestTimingType.RightBeforeWithering)
+                    && (this.__internal__getTimeUntilStage(plot, PlotStage.Berry) > 15))
+                {
+                    continue;
+                }
+                // If we are trying to unlock a new berry, dont harvest other berries so they can mutate
+                if ((this.__internal__currentStrategy.berryToUnlock !== undefined)
+                    && (plot.berryData.name !== this.__internal__currentStrategy.berryToUnlock))
```
Comment (Farigh) [inline src/lib/Farm.js:327]: Why did you switch from `plot.berry` to `plot.berryData.name` here ?

The berry name is not te same as the berry type

**Review by Farigh — CHANGES_REQUESTED** [review]
The if rework does not seem right, I've pointed a problem, but there are more to it.
For example, you did not handle `ForcePlantBerriesAsked`

File: src/lib/Farm.js (hunk 1 above)
Comment (Daniferrito) [reply src/lib/Farm.js:327]: Just copied the wrong thing. Commiting it now.

**Comment by Daniferrito** [comment]
Implemented changes as requested.

For `ForcePlantBerriesAsked`, I was assuming that it would harvest them anyways, but it is true that it would still listen to the `HarvestLate` setting. It now checks for it at the start, and harvests as soon as possible, ignoring all other settings.

We might want to change this later if even in forced mode it uses supporting plants, like Wacan, Pashto, or even Petaya. However, because I dont think forced mode is ever right, I dont think I will handle that (forced mode is for quests, and some of the quests for berries take WAY longer (days, instead of minutes) than any other quest, so I always have all farm quests disabled from the automation)

**Review by Daniferrito — COMMENTED** [review]
@Farigh I believe the issues you mentioned have been addressed. I dont see a button to mark them as done (but I did mark the one code-change as done)

**Review by Farigh — APPROVED** [review]
Sorry for the late review, I have been quite busy the past few weeks ....

I reworked the function again, as your implementation made some changes to the initial intent.
I disabled the new setting if the 'Focus on unlocking plots and new berries' as well, to make it clear to the user that this feature is not used in this case.

Tell me if this is fine with your intended behaviour

**Comment by Daniferrito** [comment]
No problem, we all have other priorities.

Disabling the option if the setting is 'Focus on unlocking plots and new berries' is indeed the intended behaviour.

**Comment by Farigh** [comment]
Sorry I should not have been too vague.
I meant: could you make sure that the behaviour still matches your expectation ?
I tried it using the Cherry berries and it does wait until the berry time to death was under 15 before harvesting. But I'm not sure it's the right way to test it.

**Comment by Daniferrito** [comment]
I will, but it sounds fine. The only thing that could be wrong is if you reverted the check for growth speed.

I will report back with the tests tomorrow, as it is a bit late now.

**Comment by Daniferrito** [comment]
It is working as intended, at least with the setting on and off.

I have not tested the unlocking of new berries.

**Comment by Farigh** [comment]
I tested the focus by insuring the unwanted crops were harvested as soon as possible and that planting strategy worked properly.
The other parts of this mechanic should be guaranteed by unit tests.

**Comment by Farigh** [comment]
Thanks again for your contributions ;)

**Pull request merged**